2 book listing

shalakashelly opened the pull request.

shalakashelly committed:
Task 1 - fake login/post-login pages

Fix for review comments

shalakashelly committed:
Task 1 - fake login/post-login pages

Fix for review comments

shalakashelly committed:
Task 1 - fake login/post-login pages

Fix for comments

shalakashelly committed:
Task 2 - book listing

shalakashelly committed:
Merging commit

shalakashelly committed:
Introduced Books.tsx as a page to render BookCard component

phistuck commented on src/App.tsx:
```diff
@@ -26,15 +26,17 @@ function App() {
-        </div>
-        <button name="submit" type="submit" disabled={isDisabled}>Login</button>
-      </form>
+    <div className="form-center">
```
1. I think the popular naming convention for classes is now ["BEM"](https://getbem.com/), are you going to follow some convention?
2. The names of the classes sound either presentational (`form-center` should always be centred), or tied to the CSS rules (`flex-container` would use flexbox), so if you want to change it to something else in the future, you will have to change the class name, is it forward compatible enough for you?
3. It would be good to use CSS (OR SCSS) modules, which try to guarantee isolation and prevent style leaks
CSS/SCSS modules are usually utilised by naming the CSS/SCSS file .module.css and importing the classes from its default export

So -
```
import styles from './Book.module.scss';

<div className={styles.Book}>I am styled by the .Book selector in the module</div>
```

That does not mean you cannot have global (non-module) CSS/SCSS as well. You can combine both of them, but you should only import the global CSS/SCSS from App.tsx (or whatever it is that wraps the entire application/all of the pages) - other files should use modules exclusively

phistuck commented on src/components/BookCard.tsx:
```diff
@@ -0,0 +1,17 @@
+const BookCard = (key: any, book: IBook) => {
+
+    return (
+        <div key={key} className="book">
```
Nit - `key` is not a real prop that you should read and it is unnecessary in this context, it should only be used when returning an array of elements. This component returns a single element

phistuck commented on src/components/Home.test.tsx:
```diff
@@ -6,6 +6,8 @@ import { authUser } from '../store/userSlice';
 
 const mockStore = configureMockStore();
 let initialState = {};
+let store: MockStore;
```
Nit - is `let store: MockStore;` needed at all?

I retract this part -
> Same for `let initialState = {};` (you removed the `let` from line 40 (38 in the before version, if you put it back (but hopefully use `const` instead), this one should not be needed)

phistuck commented on src/components/Home.tsx:
```diff
@@ -9,8 +10,12 @@ export default function Home() {
-        <div>
-            {user ? <p>Welcome, {user.username}!</p> : <p>Please log in</p>}
+        <div className="main-container">
+            {user ? 
+                <div className="grid-wrapper">
+                    <p className="grid-title">Welcome, {user.username}!</p>
+                    <Books />
+                </div> : <p className="grid-title">Please log in</p>}
```
Nit - this ternary is not very easy to read (especially with this kind of formatting/lack of new lines/indentation), maybe put stuff in variables and reverse the order (almost like an early bail)

phistuck commented on src/interfaces/books.ts:
```diff
@@ -0,0 +1,14 @@
+    title: string,
+    author: string,
+    publisher?: string,
+    publication_date?: string,
```
Nit - it would be good to choose a single naming convention for the project - either camelCase, or unix_hacker_style (or anything else you fancy), not both of them
Nit - should this be typed as a `Date` instead?

shalakashelly replied:
Changing it to camelCase.
For Date, the JSON has year format, for e.g., "1954", so I'll prefer to keep it as string, as it's less complicated.

phistuck replied:
Well, I guess it is not really a date, it is just a year, so maybe it should be a number
https://stackoverflow.com/a/25126647

shalakashelly replied:
Date also has something like this, "8th century BCE". So I suppose this should remain a string?

phistuck replied:
Oy. :| We can revisit, yuck

phistuck commented on src/interfaces/books.ts:
```diff
@@ -0,0 +1,14 @@
+    publication_date?: string,
+    isbn?: string,
+    description: string,
+    cover_image: string,
```
Nit - should there be a fancier image data structure? One that has the `width` and `height`, maybe also an `alt`, for example?

shalakashelly replied:
Sure, we can add this in the upcoming issues, while detailing the layout of books. I'll prefer to keep it as it is, for now.

phistuck commented on src/interfaces/books.ts:
```diff
@@ -0,0 +1,14 @@
+    description: string,
+    cover_image: string,
+    categories: Category[],
+    tags: Tags[]
```
Nit - does it make sense for the frontend to hard code the category and tag options?

phistuck commented on src/pages/books.tsx:
```diff
@@ -0,0 +1,39 @@
+import { IBook } from '../interfaces/books';
+
+const Books = () => {
+    const [books, setBooks] = useState<IBook[]>([]);
+    const [isLoading, setIsLoading] = useState(true);
```
Just sharing - a little trick I have been using is to have a single tristate value here -
- `undefined` (as the initial value) when loading the data
- `null` when loading fails
- The actual data (`IBook[]`) when loading succeeds

It is fine to use two `useState`, just sharing what I have been doing instead

phistuck commented on src/pages/books.tsx:
```diff
@@ -0,0 +1,39 @@
+
+    useEffect(() => {
+        fetchBookData();
+    }, []);
```
Nit - ESList would/should generally be complaining about the lack of `fetchBookData` in the dependency list of `useEffect`. While it is annoying, it should ideally be listened to

The rules of hooks says you have to put anything you use in the hook in the dependency list
Exceptions are `setFoo`s from `useState` and the returned value of `useRef`

phistuck commented on src/pages/books.tsx:
```diff
@@ -0,0 +1,39 @@
+            const json = await response.json();
+            setBooks(json);
+        } catch (error) {
+            console.error('Error fetching data:', error);
```
Are we fine with having a page with no content/message in case there is an error?

shalakashelly replied:
Yes, good for now

phistuck commented on src/pages/books.tsx:
```diff
@@ -0,0 +1,39 @@
+
+    const fetchBookData = async () => {
+        try {
+            const response = await fetch('/mockData/books.json');
+            const json = await response.json();
```
Nit - it would be good to have a simple abstraction of this, maybe create a `fetchJSON` function that will do that instead of potentially doing this every time you need to fetch data
Nit - ideally, there should be a service for fetching data instead of hard coding URLs in components

phistuck commented on src/pages/books.tsx:
Nit - the dashboard/home seems to also be a page, but it is not under pages/
Same for the login page

shalakashelly replied:
I don't want the home to be a page. I want it to be a component.

phistuck replied:
Though it is a page
By the way, books is not a page at all, it is a component. `App` and `Home` are currently the only pages because only they are what URLs can directly render, `Home` renders `Books`

phistuck commented on src/components/BookCard.tsx:
```diff
@@ -0,0 +1,17 @@
+    return (
+        <div key={key} className="book">
+            <img src="../images/book.jpg"
+                alt={book.title}
```
Nit - if the image does not have a meaningful alt and it is just repeating something that already exists in elsewhere, the blind user will hear it twice/more. You can maybe use `${book.title} cover}`, though I am not sure how helpful that would be to the blind user. To mark an image as presentational (and so screen readers would ignore it), you can use an empty string alt - `alt=""`

phistuck reviewed (changes requested):
Missing functionality -
- 17 books instead of the requested 15 (naa, joking)
- Rating on each book
- An "already read" button on each book

phistuck commented:
~What about tests?~

shalakashelly committed:
Fixed for review comments.

phistuck commented on src/components/BookCard.tsx:
```diff
@@ -0,0 +1,20 @@
+import { TBook } from '../types/books';
+
+interface BookCardProps {
+    id?: number;
```
Nit - this is not being passed in or used, can probably be removed (the ID will ideally be in `book` already)

shalakashelly committed:
- Renaming Home to Dashboard, and moving it to pages.
- Moving books and BookCard to components.
- Test cases will be added in the upcoming commits.

shalakashelly committed:
- Removing hardcoded values for Catrgories and Types.
- camelCase for publication_date and cover_image.

shalakashelly committed:
Used single tristate value for books.
- undefined (as the initial value) when loading the data
- null when loading fails
- The actual data (TBook[]) when loading succeeds

shalakashelly committed:
Changing the data type of publicationDate back to string.

shalakashelly committed:
Fix - Changed the way the Dashboard page was rendered.

shalakashelly committed:
Fix - Moved the fetch to services.

shalakashelly committed:
Fix - Using async/await to fetch books. Provided return type to fetchAllBooks function.

phistuck commented on src/pages/Dashboard/Dashboard.tsx:
```diff
@@ -0,0 +1,26 @@
+                </div>
+            );
+        } else {
+            <p className="grid-title">Please log in</p>
```
Since it is not returned, I think it is not going to be rendered

shalakashelly committed:
Fix - Added missing return to render 'Please log in'

shalakashelly committed:
Fix - Used BEM methodology for CSS

shalakashelly committed:
- Used localStorage to retrive username, to avoid going back to the log in page when the dashboard is refreshed.

shalakashelly committed:
- Created a custom hook for fetching books.

shalakashelly committed:
- Added tests for BookCard and Book component.
- Added SASS modules for CSS.
- Added a logger for console and error statements.

shalakashelly committed:
- Declared global variable in testUtils

shalakashelly committed:
- Changed type of prev to boolean in testUtils

shalakashelly committed:
- Removing comments

phistuck commented on src/hooks/useBooks.tsx:
This can be removed, not used

phistuck commented on src/components/Books/books.tsx:
Nit - this should be renamed to "Books.tsx". If you did that and it did not work, maybe first commit a rename to "books-component.tsx" and then again to "Books.tsx", otherwise just changing the case of the letters might not be reflected

phistuck commented on src/services/CONSTANTS.js:
Nit - this should be renamed to "constants.tsx". If you did that and it did not work, maybe first commit a rename to "constants-stuff.ts" and then again to "constants.ts", otherwise just changing the case of the letters might not be reflected
